training_record/services.py

```diff
@@ -79,6 +79,63 @@ def create_off_campus_record_from_raw_data(
                 )
         return record
 
+    # pylint: disable=invalid-name
+    # pylint: disable=redefined-builtin
+    @staticmethod
+    def update_off_campus_record_from_raw_data(
+            record, off_campus_event=None, user=None,
+            contents=None, attachments=None):
+        '''Update the off-campus record'''
+        if off_campus_event is None:
+            raise BadRequest('校外培训活动数据格式无效')
+        if user is None or not isinstance(user, User):
+            raise BadRequest('用户无效')
+        if contents is None:
+            contents = []
+        if attachments is None:
+            attachments = []
+
+        with transaction.atomic():
+            # get the record to be updated
+            try:
+                record = Record.objects.get(id=record.id)
+            except Exception:
+                raise BadRequest('校外培训记录无效')
+
+            # update the offCampusEvent
+            try:
+                original_off_campus_event = OffCampusEvent.objects.get(
```

Comment:
You might need to use `.select_for_update()` to lock this instance until the end of the transaction.